backstory/src/test/java/com/google/sps/servlets/AnalyzeImageServletTest.java

```diff
@@ -0,0 +1,178 @@
+import com.google.appengine.api.users.User;
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.runners.MockitoJUnitRunner;
+import org.mockito.ArgumentCaptor;
+import java.util.ArrayList;
+import java.util.List;
+import java.io.IOException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.Map;
+import java.util.HashMap;
+import com.google.sps.perspective.data.APINotAvailableException;
+import com.google.sps.perspective.data.NoAppropriateStoryException;
+import com.google.sps.perspective.data.StoryDecision;
+
+/**
+ * Tests for the analyze image servlet, which contains all image analysis and backstory generation
+ * functionalities.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public final class AnalyzeImageServletTest {
+
+  /**
+   * Tests that the correct input is passed from the initial call to the managers, 
+   * and that the correct ouput is passed from the managers to the permanent storage service.
+   */
+  @Test
+  public void testDoPostInputAndOuput() throws APINotAvailableException, NoAppropriateStoryException, IOException {
+    HttpServletRequest mockRequest = mock(HttpServletRequest.class);
+    HttpServletResponse mockResponse = mock(HttpServletResponse.class);
+
+    // 
+    AnalyzeImageServlet servlet = new AnalyzeImageServlet();
+    ImagesManager mockImagesManager = mock(VisionImagesManager.class);
+    StoryManager mockStoryManager = mock(StoryManagerImpl.class);
+    StoryAnalysisManager mockStoryAnalysisManager = mock(PerspectiveStoryAnalysisManager.class);
+    DatastoreService mockDatastoreService = mock(DatastoreService.class);
+    Entity mockAnalyzedImageEntity = mock(Entity.class);
+    BlobstoreManager mockBlobstoreManager = mock(BlobstoreManager.class);
+    UserService mockUserService = mock(UserService.class);
+
+    String userEmail = "user@gmail.com";
+    String userAuthenticationDomain = "authentication";
+    User testUser = new User(userEmail, userAuthenticationDomain);
+
+
+    // Creates the required real test objects, and sets the required behavior of the mocks.
+    boolean userIsLoggedIn = true;
+    when(mockUserService.isUserLoggedIn()).thenReturn(userIsLoggedIn);
+    when(mockUserService.getCurrentUser()).thenReturn(testUser);
+
+    String blobKeyString = "blobKeyString";
+    when(mockBlobstoreManager.getUploadedFileBlobKeyString(any(HttpServletRequest.class), anyString()))
+        .thenReturn(blobKeyString);
+    int bytesInUploadedImage = 10;
+    byte[] uploadedImageBytes = new byte[bytesInUploadedImage];
+    when(mockBlobstoreManager.getBlobBytes(any(HttpServletRequest.class), anyString()))
+        .thenReturn(uploadedImageBytes);
+
+    String sampleDescription = "sampleDescription";
+    List<String> uploadedImageDescriptions = new ArrayList<>();
+    uploadedImageDescriptions.add(sampleDescription);
+    AnnotatedImage uploadedAnnotatedImage = mock(AnnotatedImage.class);
```

Comment:
Is it possible to create a real instance of this?

Reply:
One of the fields in AnnotatedImage cannot be instantiated without network calls, so the class would have to be rewritten to allow for mock injection - since this is a wrapper class, creating a mock for it may be the simplest solution.